README
                     Owl Platform Common Library for Java

Version 1.0.4-SNAPSHOT
Last updated October 17, 2012

Owl Platform: <http://www.owlplatform.com>
Project Website: <https://github.com/OwlPlatform/java-owl-common>

Copyright (C) 2012 Robert Moore and the Owl Platform

  This library is free software; you can redistribute it and/or modify it
  under the terms of the GNU Lesser General Public License as published by the
  Free Software Foundation; either version 2.1 of the License, or (at your
  option) any later version.

  This library is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
  for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with this library; if not, write to the Free Software Foundation,
  Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA

== About ==
  This library contains common classes and routines used by other components
  of the Owl Platform.

== Compiling ==
  This library should be compiled using the Apache Maven project management
  tool.  The project is currently compatible with Apache Maven version 3,
  which can be downloaded for free at <http://maven.apache.org/>.  To build
  the JAR file, the following command should be run from the project root
  (where the pom.xml file is located):

    mvn clean package -U

  If everything compiles correctly, then near the end of the Maven output,
  this line should appear:

    [INFO] BUILD SUCCESS

  In this case, the JAR file will be located in the ``target'' subdirectory.
  If not, please visit the project website listed at the top of this document
  for support.



CHANGELOG
                     Owl Platform Common Library for Java
                                  Change Log

Owl Platform: <http://www.owlplatform.com>
Project Website: <https://github.com/OwlPlatform/java-owl-common>

Version 1.0.4 - TBD

version 1.0.3 - October 17, 2012
 + Upgraded to SLF4J 1.6.6.

Version 1.0.2 - June 18, 2012
 + Added hexadecimal string to byte[] converter in NumericUtils.
 + Prohibiting the use of the "ALL" PHY identifier in sample messages.
 + Adding a specific "Undefined" PHY identifier to SampleMessage.

Version 1.0.1 - June 13, 2012
 + Initial version ported from GRAIL RTLS at SourceForge.net.
   <http://www.sf.net/projects/grailrtls/>


src/test/java/com/owlplatform/common/util/PairTest.java
/*
 * Owl Platform Common Library for Java
 * Copyright (C) 2012 Robert Moore and the Owl Platform
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 *  
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *  
 * You should have received a copy of the GNU Lesser General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */
package com.owlplatform.common.util;

import org.junit.Assert;
import org.junit.Test;

/**
 * Unit tests for the Pair class.
 * @author Robert Moore
 *
 */
public class PairTest {

  /**
   * An Integer value for testing.
   */
  private static final Integer INT_VALUE1 = Integer.valueOf(1);
  /**
   * An Integer value for testing.
   */
  private static final Integer INT_VALUE2 = Integer.valueOf(2);

  /**
   * A String value for testing.
   */
  private static final String STRING_VALUE1 = "One";
  /**
   * A String value for testing.
   */
  private static final String STRING_VALUE2 = "Two";
  
  /**
   * Expected string representation for INT_VALUE1 and STRING_VALUE1.
   */
  private static final String INT1_STR1_AS_STRING = "(1, One)";
  /**
   * Expected string representation for INT_VALUE2 and STRING_VALUE2.
   */
  private static final String STR2_INT2_AS_STRING = "(Two, 2)";

  /**
   * Tests value 1.
   */
  @Test
  public void testValue1() {
    Pair<String,Integer> pair = new Pair<String, Integer>();
    pair.setValue1(STRING_VALUE1);
    Assert.assertEquals(STRING_VALUE1, pair.getValue1());
    
  }

  /**
   * Tests value 2.
   */
  @Test
  public void testValue2() {
    Pair<String,Integer> pair = new Pair<String,Integer>();
    pair.setValue2(INT_VALUE1);
    Assert.assertEquals(INT_VALUE1,pair.getValue2());
  }

  /**
   * Tests conversion to a String.
   */
  @Test
  public void testToString() {
    Pair<Integer,String> pair1 = new Pair<Integer,String>();
    pair1.setValue1(INT_VALUE1);
    pair1.setValue2(STRING_VALUE1);
    Pair<String,Integer> pair2 = new Pair<String,Integer>();
    pair2.setValue1(STRING_VALUE2);
    pair2.setValue2(INT_VALUE2);
    
    Assert.assertEquals(INT1_STR1_AS_STRING,pair1.toString());
    Assert.assertEquals(STR2_INT2_AS_STRING,pair2.toString());
  }

}


src/test/java/com/owlplatform/common/SampleMessageTest.java
/*
 * Owl Platform Common Library for Java
 * Copyright (C) 2012 Robert Moore and the Owl Platform
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 *  
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *  
 * You should have received a copy of the GNU Lesser General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

package com.owlplatform.common;

import java.util.Arrays;

import junit.framework.Assert;

import org.junit.Before;
import org.junit.Test;

import com.owlplatform.common.SampleMessage;

/**
 * Test Unit class for the {@code com.owlplatform.common.SampleMessage} class.
 * 
 * @author Robert Moore
 * 
 */
public class SampleMessageTest {

  /**
   * A SampleMessage object to use during testing.
   */
  private SampleMessage basicSample;

  /**
   * A SampleMessage object to use during testing.
   */
  private SampleMessage testSample;

  /**
   * Example unknown physical layer type.
   */
  private static final byte PHY_UNKNOWN = (byte) 0xFF;

  /**
   * The expected length of a Sensor-Aggregator SampleMessage object with no
   * data.
   */
  private static final int BASE_LENGTH = 45;

  /**
   * Sample data {@code byte[]} with length 0.
   */
  private static final byte[] TEST_DATA_0BYTE = new byte[] {};

  /**
   * Sample 4-byte data {@code byte[]}.
   */
  private static final byte[] TEST_DATA_4BYTE = new byte[] { 0, 1, 2, 3 };

  /**
   * Sample 8-byte data {@code byte[]}.
   */
  private static final byte[] TEST_DATA_8BYTE = new byte[] { (byte) 0xCA, 0x11,
      (byte) 0xAB, 0x1E, (byte) 0xB0, 0x1D, (byte) 0xFA, (byte) 0xCE };

  /**
   * Sample 24-byte data {@code byte[]}.
   */
  private static final byte[] TEST_DATA_24BYTE = new byte[] { (byte) 0xCA,
      0x11, (byte) 0xAB, 0x1E, (byte) 0xB0, 0x1D, (byte) 0xFA, (byte) 0xCE,
      (byte) 0xCA, 0x11, (byte) 0xAB, 0x1E, (byte) 0xB0, 0x1D, (byte) 0xFA,
      (byte) 0xCE, (byte) 0xCA, 0x11, (byte) 0xAB, 0x1E, (byte) 0xB0, 0x1D,
      (byte) 0xFA, (byte) 0xCE };

  /**
   * Sample RSSI value.
   */
  private static final float RSSI1 = -50f;
  /**
   * Sample RSSI value.
   */
  private static final float RSSI2 = -99f;

  /**
   * Sample RSSI value.
   */
  private static final float RSSI3 = 100f;

  /**
   * A timestamp at a fixed point in time.
   */
  private static final long TIMESTAMP_FIXED = -500l;

  /**
   * A timestamp that varies with the time the test is run.
   */
  private static final long TIMESTAMP_NOW = System.currentTimeMillis();

  /**
   * A sample device id value.
   */
  private static final byte[] TEST_DEVICE_ID1 = new byte[] { 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0x12, 0x34 };

  /**
   * A sample device id value.
   */
  private static final byte[] TEST_DEVICE_ID2 = new byte[] { 0, 1, 2, 3, 4, 5,
      6, 7, 8, 9, 0xA, 0xB, 0xC, 0xD, 0xE, 0xF };

  /**
   * The expected result from a call to {@code toString()} on a Sample with
   * default values.
   */
  private static final String TO_STRING_PLAIN = "Sample (-1, 0x, 0x): 0.0 @ 0";

  /**
   * The expected result from a call to {@code toString()} on a Sample with some
   * of the test data values.
   */
  private static final String TO_STRING_DATA = "Sample (2, 0x00000000000000000000000000001234, 0x000102030405060708090A0B0C0D0E0F): -50.0 @ -500 [0x00010203]";

  /**
   * Creates the two samples used for testing.
   */
  @Before
  public void createSamples() {
    this.basicSample = new SampleMessage();
    this.testSample = SampleMessage.getTestMessage();
  }

  /**
   * Tests that the physical layer value is correctly handled by the
   * SampleMessage class.
   */
  @Test
  public void testPhysicalLayer() {
    Assert.assertEquals(SampleMessage.PHYSICAL_LAYER_UNDEFINED, this.basicSample.getPhysicalLayer());

    this.basicSample.setPhysicalLayer(SampleMessage.PHYSICAL_LAYER_PIPSQUEAK);
    Assert.assertEquals(SampleMessage.PHYSICAL_LAYER_PIPSQUEAK,
        this.basicSample.getPhysicalLayer());

    this.basicSample.setPhysicalLayer(SampleMessage.PHYSICAL_LAYER_WIFI);
    Assert.assertEquals(SampleMessage.PHYSICAL_LAYER_WIFI,
        this.basicSample.getPhysicalLayer());

    this.basicSample.setPhysicalLayer(PHY_UNKNOWN);
    Assert.assertEquals(PHY_UNKNOWN, this.basicSample.getPhysicalLayer());
  }

  /**
   * Test to ensure that using {@link SampleMessage#PHYSICAL_LAYER_ALL} throws
   * an {@code IllegalArgumentException}.
   */
  @Test(expected=IllegalArgumentException.class)
  public void testInvalidPhysicalLayer() {
    this.basicSample.setPhysicalLayer(SampleMessage.PHYSICAL_LAYER_ALL);
  }

  /**
   * Tests that the length prefix values are correctly returned by the
   * SampleMessage class.
   */
  @Test
  public void testLengthPrefix() {
    // Default sensor->aggregator length is 45 bytes
    Assert.assertEquals(BASE_LENGTH, this.basicSample.getLengthPrefixSensor());
    // +1 byte for the message type
    Assert.assertEquals(BASE_LENGTH + 1,
        this.basicSample.getLengthPrefixSolver());

    this.basicSample.setSensedData(TEST_DATA_4BYTE);
    Assert.assertEquals(BASE_LENGTH + 4,
        this.basicSample.getLengthPrefixSensor());
    Assert.assertEquals(BASE_LENGTH + 5,
        this.basicSample.getLengthPrefixSolver());

    this.basicSample.setSensedData(TEST_DATA_8BYTE);
    Assert.assertEquals(BASE_LENGTH + 8,
        this.basicSample.getLengthPrefixSensor());
    Assert.assertEquals(BASE_LENGTH + 9,
        this.basicSample.getLengthPrefixSolver());

    this.basicSample.setSensedData(TEST_DATA_0BYTE);
    Assert.assertEquals(BASE_LENGTH, this.basicSample.getLengthPrefixSensor());
    Assert.assertEquals(BASE_LENGTH + 1,
        this.basicSample.getLengthPrefixSolver());

    this.basicSample.setSensedData(null);
    Assert.assertEquals(BASE_LENGTH, this.basicSample.getLengthPrefixSensor());
    Assert.assertEquals(BASE_LENGTH + 1,
        this.basicSample.getLengthPrefixSolver());
  }

  /**
   * Tests that the device ID values are handled correctly within SampleMessage.
   */
  @Test
  public void testDeviceId() {
    Assert.assertNull(this.basicSample.getDeviceId());

    this.basicSample.setDeviceId(TEST_DEVICE_ID1);
    Assert.assertTrue(Arrays.equals(TEST_DEVICE_ID1,
        this.basicSample.getDeviceId()));

    this.basicSample.setDeviceId(TEST_DEVICE_ID2);
    Assert.assertTrue(Arrays.equals(TEST_DEVICE_ID2,
        this.basicSample.getDeviceId()));
  }

  /**
   * Tests that setting a null device ID will throw a RuntimeException.
   */
  @Test(expected = RuntimeException.class)
  public void testDeviceIdNull() {
    this.basicSample.setDeviceId(null);
  }

  /**
   * Tests that setting a device ID shorter than 16 bytes will throw a
   * RuntimeException.
   */
  @Test(expected = RuntimeException.class)
  public void testDeviceIdShort() {
    this.basicSample.setDeviceId(TEST_DATA_4BYTE);
  }

  /**
   * Tests that setting a device ID longer than 16 bytes will throw a
   * RuntimeException.
   */
  @Test(expected = RuntimeException.class)
  public void testDeviceIdLong() {
    this.basicSample.setDeviceId(TEST_DATA_24BYTE);
  }

  /**
   * Tests to ensure that receiver ID values are treated correctly within the
   * SampleMessage class.
   */
  @Test
  public void testReceiverId() {
    Assert.assertNull(this.basicSample.getReceiverId());

    this.basicSample.setReceiverId(TEST_DEVICE_ID1);
    Assert.assertTrue(Arrays.equals(TEST_DEVICE_ID1,
        this.basicSample.getReceiverId()));

    this.basicSample.setReceiverId(TEST_DEVICE_ID2);
    Assert.assertTrue(Arrays.equals(TEST_DEVICE_ID2,
        this.basicSample.getReceiverId()));
  }

  /**
   * Ensures that a null receiver ID will throw a RuntimeException.
   */
  @Test(expected = RuntimeException.class)
  public void testReceiverIdNull() {
    this.basicSample.setReceiverId(null);
  }

  /**
   * Ensures that setting a receiver ID shorter than 16 bytes will throw a
   * RuntimeException.
   */
  @Test(expected = RuntimeException.class)
  public void testReceiverIdShort() {
    this.basicSample.setReceiverId(TEST_DATA_4BYTE);
  }

  /**
   * Ensures that setting a receiver ID long than 16 bytes will throw a
   * RuntimeException.
   */
  @Test(expected = RuntimeException.class)
  public void testReceiverIdLong() {
    this.basicSample.setReceiverId(TEST_DATA_24BYTE);
  }

  /**
   * Tests to ensure that sample timestamps are set correctly.
   */
  @Test
  public void testReceiverTimeStamp() {
    Assert.assertEquals(0l, this.basicSample.getReceiverTimeStamp());

    this.basicSample.setReceiverTimeStamp(TIMESTAMP_FIXED);
    Assert.assertEquals(TIMESTAMP_FIXED,
        this.basicSample.getReceiverTimeStamp());

    this.basicSample.setReceiverTimeStamp(TIMESTAMP_NOW);
    Assert.assertEquals(TIMESTAMP_NOW, this.basicSample.getReceiverTimeStamp());
  }

  /**
   * Tests to ensure that RSSI values are handled correctly.
   */
  @Test
  public void testRssi() {
    Assert.assertEquals(0f, this.basicSample.getRssi(), 0.001f);

    this.basicSample.setRssi(RSSI1);
    Assert.assertEquals(RSSI1, this.basicSample.getRssi(), 0.001f);

    this.basicSample.setRssi(RSSI2);
    Assert.assertEquals(RSSI2, this.basicSample.getRssi(), 0.001f);

    this.basicSample.setRssi(RSSI3);
    Assert.assertEquals(RSSI3, this.basicSample.getRssi(), 0.001f);
  }

  /**
   * Tests to ensure that sensed data values are handled correctly.
   */
  @Test
  public void testSensedData() {
    Assert.assertNull(this.basicSample.getSensedData());

    this.basicSample.setSensedData(TEST_DATA_0BYTE);
    Assert.assertTrue(Arrays.equals(TEST_DATA_0BYTE,
        this.basicSample.getSensedData()));

    this.basicSample.setSensedData(TEST_DATA_4BYTE);
    Assert.assertTrue(Arrays.equals(TEST_DATA_4BYTE,
        this.basicSample.getSensedData()));

    this.basicSample.setSensedData(TEST_DATA_8BYTE);
    Assert.assertTrue(Arrays.equals(TEST_DATA_8BYTE,
        this.basicSample.getSensedData()));
  }

  /**
   * Tests the clone method to ensure that cloned objects are appropriately
   * equivalent.
   * 
   * @throws CloneNotSupportedException
   */
  @Test
  public void testClone() throws CloneNotSupportedException {
    SampleMessage clone = this.basicSample.clone();

    Assert.assertEquals(this.basicSample.getLengthPrefixSensor(),
        clone.getLengthPrefixSensor());
    Assert.assertEquals(this.basicSample.getLengthPrefixSolver(),
        clone.getLengthPrefixSolver());
    Assert.assertEquals(this.basicSample.getCreationTimestamp(),
        clone.getCreationTimestamp());
    Assert.assertEquals(this.basicSample.getPhysicalLayer(),
        clone.getPhysicalLayer());
    Assert.assertEquals(this.basicSample.getReceiverTimeStamp(),
        clone.getReceiverTimeStamp());
    Assert.assertEquals(this.basicSample.getRssi(), clone.getRssi(), 0.001f);
    Assert.assertTrue(Arrays.equals(this.basicSample.getDeviceId(),
        clone.getDeviceId()));
    Assert.assertTrue(Arrays.equals(this.basicSample.getReceiverId(),
        clone.getReceiverId()));
    Assert.assertTrue(Arrays.equals(this.basicSample.getSensedData(),
        clone.getSensedData()));

    this.testSample.setSensedData(TEST_DATA_4BYTE);
    clone = this.testSample.clone();

    Assert.assertEquals(this.testSample.getLengthPrefixSensor(),
        clone.getLengthPrefixSensor());
    Assert.assertEquals(this.testSample.getLengthPrefixSolver(),
        clone.getLengthPrefixSolver());
    Assert.assertEquals(this.testSample.getCreationTimestamp(),
        clone.getCreationTimestamp());
    Assert.assertEquals(this.testSample.getPhysicalLayer(),
        clone.getPhysicalLayer());
    Assert.assertEquals(this.testSample.getReceiverTimeStamp(),
        clone.getReceiverTimeStamp());
    Assert.assertEquals(this.testSample.getRssi(), clone.getRssi(), 0.001f);
    Assert.assertTrue(Arrays.equals(this.testSample.getDeviceId(),
        clone.getDeviceId()));
    Assert.assertTrue(Arrays.equals(this.testSample.getReceiverId(),
        clone.getReceiverId()));
    Assert.assertTrue(Arrays.equals(this.testSample.getSensedData(),
        clone.getSensedData()));
  }

  /**
   * Tests the toString() method of the SampleMessage class.
   */
  @Test
  public void testToString() {
    Assert.assertEquals(TO_STRING_PLAIN, this.basicSample.toString());

    this.basicSample.setDeviceId(TEST_DEVICE_ID1);
    this.basicSample.setReceiverId(TEST_DEVICE_ID2);
    this.basicSample.setRssi(RSSI1);
    this.basicSample.setPhysicalLayer(SampleMessage.PHYSICAL_LAYER_WIFI);
    this.basicSample.setReceiverTimeStamp(TIMESTAMP_FIXED);
    this.basicSample.setSensedData(TEST_DATA_4BYTE);

    Assert.assertEquals(TO_STRING_DATA, this.basicSample.toString());
  }

}


src/test/java/com/owlplatform/common/util/LRUCacheTest.java
/*
 * Owl Platform Common Library for Java
 * Copyright (C) 2012 Robert Moore and the Owl Platform
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 *  
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *  
 * You should have received a copy of the GNU Lesser General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */
package com.owlplatform.common.util;

import java.util.Iterator;

import junit.framework.Assert;

import org.junit.Test;

/**
 * Unit test for the LRUCache class.
 * @author Robert Moore
 *
 */
public class LRUCacheTest {

  /**
   * Set of Integer keys to use in the cache.
   */
  private static final Integer[] keys = new Integer[] { Integer.valueOf(0),
      Integer.valueOf(2), Integer.valueOf(1), Integer.valueOf(4),
      Integer.valueOf(3), Integer.valueOf(6), Integer.valueOf(5),
      Integer.valueOf(8) };
  /**
   * A set of String values to use in the cache.
   */
  private static final String[] values = new String[] { "Zero", "Two", "One",
      "Four", "Three", "Six", "Five", "Eight" };

  /**
   * Tests the LRUCache class.
   */
  @Test
  public void test() {

    // Test insertion order
    LRUCache<Integer, String> cache = new LRUCache<Integer, String>(10);
    int i;
    for (i = 0; i < keys.length; ++i) {
      cache.put(keys[i], values[i]);
    }

    i = 0;
    for (Iterator<Integer> iter = cache.keySet().iterator(); iter.hasNext();) {
      Integer key = iter.next();
      Assert.assertEquals(keys[i], key);
      ++i;
    }

    i = 0;
    for (Iterator<String> iter = cache.values().iterator(); iter.hasNext();) {
      String value = iter.next();
      Assert.assertEquals(values[i], value);
      ++i;
    }

    // Access in reverse order and check
    for (i = keys.length - 1; i >= 0; --i) {
      cache.get(keys[i]);
    }
    i = keys.length;
    for (Iterator<Integer> iter = cache.keySet().iterator(); iter.hasNext();) {
      --i;
      Integer key = iter.next();
      Assert.assertEquals(keys[i], key);
    }

    i = keys.length;
    for (Iterator<String> iter = cache.values().iterator(); iter.hasNext();) {
      --i;
      String value = iter.next();
      Assert.assertEquals(values[i], value);
    }

    // Access in forward order and check
    for (i = 0; i < keys.length; ++i) {
      cache.get(keys[i]);
    }
    i = 0;
    for (Iterator<Integer> iter = cache.keySet().iterator(); iter.hasNext();) {
      Integer key = iter.next();
      Assert.assertEquals(keys[i], key);
      ++i;
    }

    i = 0;
    for (Iterator<String> iter = cache.values().iterator(); iter.hasNext();) {
      String value = iter.next();
      Assert.assertEquals(values[i], value);
      ++i;
    }
  }
}


src/main/java/com/owlplatform/common/SampleMessage.java
/*
 * Owl Platform Common Library for Java
 * Copyright (C) 2012 Robert Moore and the Owl Platform
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 *  
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *  
 * You should have received a copy of the GNU Lesser General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

package com.owlplatform.common;

import java.util.Arrays;

import com.owlplatform.common.util.NumericUtils;

/**
 * Represents a sample sent to or from an Aggregator as defined in the Owl
 * Platform Aggregator-Solver and Sensor-Aggregator protocols.
 * 
 * @author Robert Moore II
 * 
 */
public class SampleMessage implements Cloneable {

  /**
   * The size, in octets, of the device identifier.
   */
  public static final int DEVICE_ID_SIZE = 16;

  /**
   * The identifier used for this message.
   */
  public static final byte MESSAGE_TYPE = 6;

  /**
   * An undefined physical layer type.
   */
  public static final byte PHYSICAL_LAYER_UNDEFINED = (byte)0xFF;
  
  /**
   * Physical layer for filtering any physical layer type. This should not be
   * used as the type of a {@code SampleMessage}.
   */
  public static final byte PHYSICAL_LAYER_ALL = 0;
  /**
   * Physical layer for Pipsqueak devices.
   */
  public static final byte PHYSICAL_LAYER_PIPSQUEAK = 1;

  /**
   * Physical layer for WiFi devices.
   */
  public static final byte PHYSICAL_LAYER_WIFI = 2;

  /**
   * Physical layer for WINS devices.
   */
  public static final byte PHYSICAL_LAYER_WINS = 3;

  /**
   * The physical layer type of the transmitter/receiver identified in this
   * message.
   */
  private byte physicalLayer = PHYSICAL_LAYER_UNDEFINED;

  /**
   * Returns the physical layer type for the transmitter/receiver identified in
   * this message.
   * 
   * @return the physical layer type for the transmitter/receiver identified in
   *         this message.
   */
  public byte getPhysicalLayer() {
    return this.physicalLayer;
  }

  /**
   * Sets the physical layer type for the transmitter/receiver identified in
   * this message.
   * 
   * @param physicalLayer
   *          the physical layer type for the transmitter/receiver identified in
   *          this message.
   * @throws IllegalArgumentException if the value of {@code physicalLayer}
   *  is equal to {@link #PHYSICAL_LAYER_ALL}.
   */
  public void setPhysicalLayer(byte physicalLayer) {
    if (physicalLayer == PHYSICAL_LAYER_ALL) {
      throw new IllegalArgumentException("Invalid physical layer type. "
          + PHYSICAL_LAYER_ALL + " is reserved for filtering only.");
    }
    this.physicalLayer = physicalLayer;
  }

  /**
   * The device identifier for the transmitter referenced in this message.
   */
  private byte[] deviceId;

  /**
   * The device identifier for the receiver referenced in this message.
   */
  private byte[] receiverId;

  /**
   * The UNIX timestamp indicating when this sample was observed by the
   * receiver.
   */
  private long receiverTimeStamp;

  /**
   * The Received Signal Strength Indicator (RSSI) for this sample.
   */
  private float rssi;

  /**
   * The time when this message object was created.
   */
  private final long creationTimestamp;

  /**
   * The raw sensed data provided by the transmitter or receiver for this
   * sample.
   */
  private byte[] sensedData = null;

  /**
   * Creates a Sample message with the current time as the timestamp. No other
   * fields are set.
   */
  public SampleMessage() {
    this.creationTimestamp = System.currentTimeMillis();
  }

  /**
   * Creates a Sample message with the specified creation timestamp. No other
   * fields are set.
   * 
   * @param timestamp
   *          the time at which this sample was created.
   */
  public SampleMessage(final long timestamp) {
    this.creationTimestamp = timestamp;
  }

  /**
   * The length of this message, in bytes, as encoded according to the
   * Sensor-Aggregator protocol.
   * 
   * @return the length of this message.
   */
  public int getLengthPrefixSensor() {
    // physicalLayer, devId, recvId, timestamp, rssi
    int length = 1 + DEVICE_ID_SIZE * 2 + 8 + 4;
    if (this.sensedData != null) {
      length += this.sensedData.length;
    }
    return length;
  }

  /**
   * The length of this message, in bytes, as encoded according to the
   * Aggregator-Solver protocol.
   * 
   * @return the length of this message.
   */
  public int getLengthPrefixSolver() {
    // physicalLayer, messageId, devId, recvId, timestamp, rssi
    int length = 2 + DEVICE_ID_SIZE * 2 + 8 + 4;
    if (this.sensedData != null) {
      length += this.sensedData.length;
    }
    return length;
  }

  /**
   * Returns the device identifier for the transmitter referenced in this
   * sample.
   * 
   * @return the device identifier for the transmitter referenced in this
   *         sample.
   */
  public byte[] getDeviceId() {
    return this.deviceId;
  }

  /**
   * Sets the device identifier for the transmitter referenced in this sample.
   * 
   * @param deviceId
   *          the device identifier for the transmitter referenced in this
   *          sample.
   */
  public void setDeviceId(byte[] deviceId) {
    if (deviceId == null) {
      throw new RuntimeException("Device ID cannot be null.");
    }
    if (deviceId.length != DEVICE_ID_SIZE) {
      throw new RuntimeException(String.format(
          "Device ID must be %d bytes long.", Integer.valueOf(DEVICE_ID_SIZE)));
    }
    this.deviceId = deviceId;
  }

  /**
   * Returns the device identifier for the receiver referenced in this sample.
   * 
   * @return the device identifier for the receiver referenced in this sample.
   */
  public byte[] getReceiverId() {
    return this.receiverId;
  }

  /**
   * Sets the device identifier for the receiver referenced in this sample.
   * 
   * @param receiverId
   *          the device identifier for the receiver referenced in this sample.
   */
  public void setReceiverId(byte[] receiverId) {
    if (receiverId == null) {
      throw new RuntimeException("Receiver ID cannot be null.");
    }
    if (receiverId.length != DEVICE_ID_SIZE) {
      throw new RuntimeException(
          String.format("Receiver ID must be %d bytes long.",
              Integer.valueOf(DEVICE_ID_SIZE)));
    }
    this.receiverId = receiverId;
  }

  /**
   * Returns the UNIX timestamp indicating when this sample was received by the
   * receiver.
   * 
   * @return the UNIX timestamp indicating when this sample was received by the
   *         receiver.
   */
  public long getReceiverTimeStamp() {
    return this.receiverTimeStamp;
  }

  /**
   * Sets the UNIX timestamp indicating when this sample was received by the
   * receiver.
   * 
   * @param receiverTimeStamp
   *          the UNIX timestamp indicating when this sample was received by the
   *          receiver.
   */
  public void setReceiverTimeStamp(long receiverTimeStamp) {
    this.receiverTimeStamp = receiverTimeStamp;
  }

  /**
   * Returns the Received Signal Strength Indicator (RSSI) observed by the
   * receiver for this sample.
   * 
   * @return the Received Signal Strength Indicator (RSSI) observed by the
   *         receiver for this sample.
   */
  public float getRssi() {
    return this.rssi;
  }

  /**
   * Sets the Received Signal Strength Indicator (RSSI) observed by the receiver
   * for this sample.
   * 
   * @param rssi
   *          the Received Signal Strength Indicator (RSSI) observed by the
   *          receiver for this sample.
   */
  public void setRssi(float rssi) {
    this.rssi = rssi;
  }

  /**
   * Returns the raw sensed data sent by the transmitter and observed by the
   * receiver for this sample.
   * 
   * @return the raw sensed data sent by the transmitter and observed by the
   *         receiver for this sample.
   */
  public byte[] getSensedData() {
    return this.sensedData;
  }

  /**
   * Sets the raw sensed data sent by the transmitter and observed by the
   * receiver for this sample.
   * 
   * @param sensedData
   *          the raw sensed data sent by the transmitter and observed by the
   *          receiver for this sample.
   */
  public void setSensedData(byte[] sensedData) {
    this.sensedData = sensedData;
  }

  @Override
  public String toString() {
    StringBuffer sb = new StringBuffer();
    sb.append("Sample (").append(this.getPhysicalLayer());
    sb.append(", ");
    sb.append(NumericUtils.toHexString(this.getDeviceId()));
    sb.append(", ");
    sb.append(NumericUtils.toHexString(this.getReceiverId()));
    sb.append("): ");
    sb.append(this.getRssi());
    sb.append(" @ ");
    sb.append(this.getReceiverTimeStamp());
    if (this.getSensedData() != null) {
      sb.append(" [").append(NumericUtils.toHexString(this.getSensedData()));
      sb.append(']');
    }

    return sb.toString();
  }

  /**
   * Returns a timestamp indicating when this Sample was created.
   * 
   * @return the creation timestamp of this Sample.
   * @see #SampleMessage()
   * @see #SampleMessage(long)
   */
  public long getCreationTimestamp() {
    return this.creationTimestamp;
  }

  /**
   * Generates a generic message with the following values.
   * <ul>
   * <li>DeviceId = 0000 0000 0000 0000 0000 0000 0000 0001</li>
   * <li>ReceiverId = 0000 0000 0000 0000 0000 0000 0000 0001</li>
   * <li>Physical layer = 01</li>
   * <li>Rssi = -50</li>
   * <li>Data = 0000</li>
   * <li>Timestamp = current time</li>
   * </ul>
   * 
   * @return the created message.
   */
  public static SampleMessage getTestMessage() {
    SampleMessage message = new SampleMessage();
    message.setDeviceId(new byte[] { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 1 });
    message.setReceiverId(new byte[] { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 1 });
    message.setPhysicalLayer((byte) 1);
    message.setRssi(-50f);
    message.setSensedData(new byte[] { 0, 0 });
    return message;
  }

  @Override
  public SampleMessage clone() throws CloneNotSupportedException {
    SampleMessage clone = (SampleMessage) super.clone();
    if (this.sensedData != null) {
      clone.setSensedData(Arrays
          .copyOf(this.sensedData, this.sensedData.length));
    }
    return clone;
  }
}
